library/choose_disk.py

```diff
@@ -0,0 +1,651 @@
+def is_ceph_disk(partition):
+    '''
+    Check if a partition is containing some ceph structures
+    '''
+    stdout = subprocess.check_output(["lsblk", "-no", "PARTLABEL", "%s" % partition])
+
+    if "ceph data" in stdout:
+        return "data"
+
+    if "ceph journal" in stdout:
+        return "journal"
+
+    return ""
+
+
+def select_only_free_devices(physical_disks):
+    '''
+    It's important reporting only devices that are not currently used.
+    '''
+    selected_devices = {}
+    logger.info('Detecting free devices')
+
+    for physical_disk in sorted(physical_disks):
+        ceph_disk = ""
+        current_physical_disk = physical_disks[physical_disk]
+
+        # Don't consider devices that doesn't support partitions
+        if 'partitions' not in current_physical_disk:
+            logger.info(' Ignoring %10s : Device doesnt support partitioning', physical_disk)
+            continue
+
+        # Don't consider the device if partition list is not empty,
+        # A disk that is already partionned may contain important data
+        # It's up to the admin to zap the device before using it in ceph-ansible
+        # There is only an exception :
+        #    if the partitions are from ceph, it surely means that we have to
+        #    reuse them
+        if len(current_physical_disk['partitions']) > 0:
+            for partition in sorted(current_physical_disk['partitions']):
+                disk_type = is_ceph_disk("/dev/" + partition)
+                if disk_type:
+                    if ceph_disk:
+                        ceph_disk += " + "
+                    # Saving the ceph type (journal and/or data)
+                    ceph_disk += disk_type
+
+            if not ceph_disk:
+                logger.info(' Ignoring %10s : Device has exisiting partitions', physical_disk)
+                continue
+
+        selected_devices[physical_disk] = physical_disks[physical_disk]
+        selected_devices[physical_disk]['bdev'] = '/dev/' + physical_disk
+
+        if ceph_disk:
+            selected_devices[physical_disk]['ceph_prepared'] = ceph_disk
+            logger.info(' Adding   %10s : Ceph disk detected (%s)', physical_disk, ceph_disk)
+        else:
+            logger.info(' Adding   %10s : %s', physical_disk, selected_devices[physical_disk]['bdev'])
+
+    return selected_devices
+
+
+def get_block_devices_persistent_name(physical_disks):
+    '''
+    Replace the short name (sda) by the persistent naming 'by-id'
+    '''
+    directory = "/dev/disk/by-id/"
+
+    logger.info('Finding persistent disks name')
+
+    # If the directory doesn't exist, reports the list as-is
+    if not os.path.isdir(directory):
+        logger.info(' Cannot open %s', directory)
+        return physical_disks
+
+    final_disks = {}
+    matching_devices = {}
+
+    # Searching in the /dev/disk/by-id/ which are the disks we look for
+    for f in os.listdir(directory):
+        device_name = os.readlink(directory + f).split("/")[-1]
```

Comment:
Python has a native way of handling traversing of paths while keeping files, dirs, and parent dirs available with `os.walk`:

    Generate the file names in a directory tree by walking the tree either top-down or bottom-up. For each directory in the tree rooted at directory top (including top itself), it yields a 3-tuple (dirpath, dirnames, filenames).

It would be preferable to just consume the native way rather than re-implementing it this way